### ISSUE TITLE
Is there a way to force CodeDeploy lifecycles to wait until a given condition is met before running?

### ISSUE THREAD

**Issue by mattejames** [awslabs/aws-codedeploy-samples#60]
When I launch a new ASG instance, I have "user data" in the Launch Configuration that, among other things, runs to ensure yum is updated on the new server.  My issue is that this user data custom bash script does not finish before the CodeDeploy lifecycle scripts run, so I end up having race conditions that prevent the new instance from deploying properly.  For example, if I download and install the latest AWS tools from S3 during the user data script, those tools are needed to register and de-register from the ELB, but if the AWS CLI tools aren't available yet, the deploy fails when it tries to re-register.

Any help is greatly appreciated!

**Comment by emmceemoore** [comment]
I'm not with the CodeDeploy team but here are some ideas to consider:
- If you're OK with the yum update happening with every deployment you could move it to a CodeDeploy script. (In our setup we have a `BeforeInstall` script which does an `apt-get update && apt-get upgrade`.)
- You could update your "user data" script to `touch` a file at the beginning and then remove it at the end. You could then add a CodeDeploy script to spin-loop until that file is gone (or until the timeout is hit).

**Comment by mattejames** [comment]
Great suggestions - thanks!

I've never made a "spin-loop" in shell, any suggested links I should check out?

**Comment by emmceemoore** [comment]
[Google it](https://www.google.com/search?q=bash+loop+while+file+exists&oq=bash+loop+while&ie=UTF-8)? The [first result](http://stackoverflow.com/a/2379904) looks promising. 😉

**Comment by mattejames** [comment]
Gotcha, yeah - sorry.  The phrase "spin-loop" caught me off guard.  Thanks for the help!

**Comment by mattejames** [comment]
One piece of data that I neglected to include in my first post was that I was using a RHEL install.  The issue was that, to my knowledge, there wasn't a way to get AWS Tools via yum with RHEL.  So, I was using my user-data script to download those tools, unzip them, and then install them because they were only needed during an initialization of the instance (not during each CodeDeploy).  However, because of the race condition issue, the AWS tools were often not installed before the CodeDeploy scripts ran, which caused my deployments to fail.  I considered installing the AWS tools into the AMI instead, but wanted to make sure I had the latest version each time and it wasn't immediately clear how to update them.

For a variety of reasons (including dramatic cost savings), we ended up switching from RHEL to Amazon Linux.  And I was pleased to find out that the AWS tools are built-in by default on those and that all I'd need to do is perform a yum update in my deployment script which I try to do during deployments anyway, so after that switch it all worked out.

Hopefully that info helps others in the future.

Thanks again, emmceemoore!
